docs/Developers/Design/Upgrade_design.md

```diff
@@ -0,0 +1,146 @@
+necessary to perform upgrades of the Gluster cluster. An administrator, either
+directly or through an upgrade service, will choose the desired version of
+Gluster (GCS) that should be running on the cluster, and the operator must
+properly and autonomously upgrade all components to match that request.
+
+# Problem overview
+
+A version of GCS is comprised of a specific set of container images plus the
+associated object manifests required to deploy them. In order to upgrade, the
+container images must be changed to those in a subsequent release. These
+updates, however, must be mindful of version compatibility between the
+components. Particularly when upgrading across a series of releases, it may be
+necessary to take intermediate upgrade steps to ensure client (CSI driver)
+versions remain compatible with server (Gluster container) versions.
+
+## Interaction with Operator Lifecycle Manager
+
+[OLM](https://github.com/operator-framework/operator-lifecycle-manager) is
+designed to manage the life cycle of operators (i.e., ensure dependencies,
+install, upgrade, etc.). It accomplishes this by maintaining a set of
+ClusterServiceVersion objects that describe the available versions of an
+operator. These objects are then assembled to create an upgrade path for a
+given operator.
+
+OLM attempts to install the latest version of an operator by walking the CSV
+versions. It does this by replacing the running operator with the next version,
+waiting for it to become ready, then repeating until the operator is
+up-to-date. This rapid-succession replacement does not have a mechanism to
+pause on a version while underlying services are updated. The implication of
+this is:
+
+- The operator is required to support upgrading its internal state from version
+  *n-1* to version *n* only. (OLM walks through each version.)
+- Every version of the operator must be capable of upgrading the Gluster
+  cluster from an arbitrarily old version to the current version. (OLM will not
+  wait for lower-level upgrades.)
+
+Because of the above issues, we must have a method for representing upgrade and
+deployment actions that can be maintained indefinitely as a part of the
+operator's logic.
+
+# Proposed solution
+
+The operator will contain a manifest of all old versions and legal version
+transitions. The operator will drive the system toward the most current (aka
+highest) version in the manifest by stepping through them. The format of the
+manifest must remain in sync with the operator code-base, but it could be either
+in the operator container or in a sidecar container within the same pod.
+
+The version sequence manifest would look something like:
+
+```yaml
+---
+
+# List of versions we know about
+- version: 5.0  # describes GCS v5
+  # The components that go into the GCS "system"
+  components:
+    - name: gluster-csi-file
+      # Template is probably a Deployment or DaemonSet, etc.
```

Comment:
what we need is below
Object/object group : `ds` or `dc` or `statefulset`
Object name: `glusterfs`
upgrade/desired Version: `GCS 2.0`

Reply:
I don't think I understand your comment. My intent is to have the yaml for the kube object embedded here (DaemonSet, Deployment, etc). That would include everything such as the image:tag and resource limits. It would also probably have some sort of templatization like jinja2. The `components` list would exist for each "version" of GCS, and the operator would always move to the highest (my semver rules) version in the file.

I'm not following the "upgrade/desired" part of your comment.